File: python/src/runapi/producer/resources/text_to_music.py
"""Producer text-to-music resource."""

from __future__ import annotations

from typing import Any, Optional

from runapi.core import Resource, RequestOptions

from ..contract_gen import CONTRACT
from ..types import CompletedTextToMusicResponse, TextToMusicResponse


class TextToMusic(Resource):
    """Generate music from exact lyrics or an instrumental brief."""

    ENDPOINT = "/api/v1/producer/text_to_music"

    RESPONSE_CLASS = TextToMusicResponse
    COMPLETED_RESPONSE_CLASS = CompletedTextToMusicResponse

    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create a task and poll until it completes."""
        task = self.create(options=options, **params)
        return self._poll_until_complete(lambda: self.get(task.id, options=options))

    def create(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create a text-to-music task and return immediately with an ``id``."""
        compacted = self._compact_params(params)
        self._validate_contract(CONTRACT["text-to-music"], compacted)
        return self._request("post", self.ENDPOINT, body=compacted, options=options)

    def get(self, id: str, options: Optional[RequestOptions] = None) -> Any:
        """Fetch the current status of a text-to-music task."""
        return self._request("get", f"{self.ENDPOINT}/{id}", options=options)

File: ruby/lib/runapi/producer/resources/text_to_music.rb
# frozen_string_literal: true

module RunApi
  module Producer
    module Resources
      # Generates music from exact lyrics or an instrumental brief.
      class TextToMusic
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/producer/text_to_music"

        RESPONSE_CLASS = Types::TextToMusicResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedTextToMusicResponse

        def initialize(http)
          @http = http
        end

        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        def create(options: nil, **params)
          params = compact_params(params)
          validate_contract!(CONTRACT["text-to-music"], params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end
      end
    end
  end
end
